.github/workflows/linux.yml

```diff
@@ -15,211 +10,36 @@ on:
-      - name: Validate cli-suite
-        run: sh validate.sh -j 2 -w ghc-8.8.4 -v --solver-benchmarks -s cli-suite
-  validate-8_6_5:
-    name: validate.sh ghc-8.6.5
-    runs-on: ubuntu-18.04
-    needs: validate-8_8_4
-    container:
-      image: phadej/ghc:8.6.5-bionic
-    steps:
-      - name: System info
-        run: |
-          uname -a
-      - name: Set PATH
-        # https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-commands-for-github-actions#adding-a-system-path
-        run: |
-          echo "$HOME/.cabal/bin" >> $GITHUB_PATH
-      - name: Install cabal-plan
-        run: |
-          mkdir -p $HOME/.cabal/bin
-          curl -sL https://github.com/haskell-hvr/cabal-plan/releases/download/v0.6.2.0/cabal-plan-0.6.2.0-x86_64-linux.xz > cabal-plan.xz
-          echo "de73600b1836d3f55e32d80385acc055fd97f60eaa0ab68a755302685f5d81bc  cabal-plan.xz" | sha256sum -c -
-          xz -d < cabal-plan.xz > $HOME/.cabal/bin/cabal-plan
-          rm -f cabal-plan.xz
-          chmod a+x $HOME/.cabal/bin/cabal-plan
-      - name: apt-get update
-        run: apt-get update
-      - name: Install `nroff` for `cabal man`
-        run: apt-get install -y groff-base
-      - name: Update Hackage index
-        run: cabal v2-update
-      # https://github.com/actions/checkout/issues/170
-      # - uses: actions/checkout@v2
-      - name: Checkout
-        run: |
-          echo $GITHUB_REF $GITHUB_SHA
-          git clone --depth 1 https://github.com/$GITHUB_REPOSITORY.git .
-          git fetch origin $GITHUB_SHA:temporary-ci-branch
-          git checkout $GITHUB_SHA || (git fetch && git checkout $GITHUB_SHA)
-      - name: Validate print-config
-        run: sh validate.sh -j 2 -w ghc-8.6.5 -v --complete-hackage-tests -s print-config
-      - name: Validate print-tool-versions
-        run: sh validate.sh -j 2 -w ghc-8.6.5 -v --complete-hackage-tests -s print-tool-versions
-      - name: Validate build
-        run: sh validate.sh -j 2 -w ghc-8.6.5 -v --complete-hackage-tests -s build
-      - name: Validate lib-tests
-        run: sh validate.sh -j 2 -w ghc-8.6.5 -v --complete-hackage-tests -s lib-tests
-      - name: Validate lib-suite
-        run: sh validate.sh -j 2 -w ghc-8.6.5 -v --complete-hackage-tests -s lib-suite
-      - name: Validate cli-tests
-        run: sh validate.sh -j 2 -w ghc-8.6.5 -v --complete-hackage-tests -s cli-tests
-      - name: Validate cli-suite
-        run: sh validate.sh -j 2 -w ghc-8.6.5 -v --complete-hackage-tests -s cli-suite
-  validate-8_4_4:
-    name: validate.sh ghc-8.4.4
+  validate:
+    strategy:
+      matrix:
+        ghc:
+          - "9.2.1"
+          - "9.0.2"
+          - "8.10.7"
+          - "8.8.4"
+          - "8.6.5"
+          - "8.4.4"
+          - "8.2.2"
+          - "8.0.2"
+          - "7.10.3"
+          - "7.8.4"
+
+    name: validate ${{ matrix.ghc }}
     runs-on: ubuntu-18.04
-    needs: validate-8_8_4
```

Comment:
This got lost, right? Is there a way to restore it? We do have high parallelism, but still, running n jobs for n ghcs before checking that the thing actually builds seems wasteful

Reply:
well, yes and no, I wasn't sure the intent of the force serialisation so I left it. We can implement either a fail-fast or a partial serialisation depending on our needs (see [docos](https://docs.github.com/en/actions/using-jobs/using-jobs-in-a-workflow#defining-prerequisite-jobs)).